File: solutions/30-42585.rb
def solution(arrangement)
  stack = ['(']
  arrangement.split(//).each_cons(2).inject(0) do |piece, (prev, curr)|
    if curr == '('
      stack.push(curr)
      piece += 0
    else
      stack.pop
      prev == ')' ? piece += 1 : piece += stack.size
    end
  end
end

p solution('()(((()())(())()))(())') #=> 17
p solution('(())') #=> 2
p solution('()') #=> 0

File: solutions/py/30-42585.py
def solution(arrangement):
    count = 0
    stack = []
    for index in range(len(arrangement)):
        if arrangement[index] == '(':
            stack.append(arrangement[index])
        else:
            stack.pop()
            if arrangement[index-1] == ')':
                count += 1
            else:
                count += len(stack)

    return count


print(solution('()(((()())(())()))(())')) # 17
